File: python/image-to-diagram/agent_state.py
import base64
from typing import TypedDict, List
from langchain.output_parsers import StructuredOutputParser
class DiagramParticipant(TypedDict):
    name: str
    shape: str
    description: str

class DiagramRelation(TypedDict):
    source: str # source
    target: str # destination
    description: str

class DiagramContainer(TypedDict):
    name: str # source
    children: List[str] # destination
    description: str

class DiagramDescription(TypedDict):
    type: str
    title: str
    participans: List[DiagramParticipant]
    relations: List[DiagramRelation]
    containers: List[DiagramContainer]
    description: str # NLP description

class AgentState(TypedDict):
    # messages: Annotated[Sequence[BaseMessage], operator.add]
    diagram_image_url_or_data: str
    diagram: DiagramDescription
    diagram_code: str

def get_image_data(image_path):

    with open(image_path, "rb") as image_file:
        image_data = image_file.read()
        base64_encoded = base64.b64encode(image_data).decode("utf-8")
        return "data:image/png;base64," + base64_encoded

File: javascript/image-to-diagram/agent_state.ts

import { z } from "zod";

const participant_schema = z.object({
  name: z.string().describe('participant name'),
  shape: z.string().describe('participant shape'), // can be an enum ?
  description: z.string().describe( 'participant description')
})

const relation_schema = z.object({
  source: z.string().describe('source participant'),
  target: z.string().describe('target participant'),
  description: z.string().describe( 'relation description')
})

const container_schema = z.object({
  name: z.string().describe('container name'),
  children: z.array( z.string() ).describe('list of contained elements name'), // can be an enum ?
  description: z.string().describe( 'container description')
})

export const diagram_schema = z.object({
  type: z.string().describe('Diagram tipology (one word). Eg. "sequence", "class", "process", etc."'),
  title: z.string().describe("Diagram summary (max one line) or title (if any)"),
  participants: z.array( participant_schema ).describe("list of participants in the diagram "),
  relations: z.array( relation_schema ).describe( "list of relations in the diagram"),
  containers: z.array( container_schema ).describe("list of participants that contain other ones in the diagram"),
  description: z.array(z.string()).describe("Step by step description of the diagram with clear indication of participants and actions between them."),

})


export interface AgentState {
    diagramImageUrlOrData?: string;
    diagramCode?: string;
    diagram?: z.infer<typeof diagram_schema>
  }
  